### PR TITLE
[VPlan] Use VPWidenIntrinsicRecipe to vp.select.

### PR DESCRIPTION
**Pull request by fhahn**
Use VPWidenIntrinsicRecipe (https://github.com/llvm/llvm-project/pull/110486)
to create vp.select intrinsics. This potentially offers an alternative
to duplicating EVL recipes for all existing recipes.

There are some recipes that will need duplicates (at least at the
moment), due to extra code-gen needs (e.g. widening loads and stores).
But in cases the intrinsic can directly be used, creating the widened
intrinsic directly would reduce the need to duplicate some recipes.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -4392,11 +4392,17 @@ void LoopVectorizationPlanner::emitInvalidCostRemarks(
...
         auto *WidenCall = dyn_cast<VPWidenCallRecipe>(R);
-        Function *CalledFn =
-            WidenCall ? WidenCall->getCalledScalarFunction()
-                      : cast<Function>(R->getOperand(R->getNumOperands() - 1)
-                                           ->getLiveInIRValue());
-        OS << " call to " << CalledFn->getName();
+        StringRef Name = "";
+        if (auto *Int = dyn_cast<VPWidenIntrinsicRecipe>(R))
+          Name = Int->getIntrinsicName();
+        else {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4398]:
```suggestion
        if (auto *Int = dyn_cast<VPWidenIntrinsicRecipe>(R)) {
          Name = Int->getIntrinsicName();
        } else {
```

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Fixed (also in PR that adds this code https://github.com/llvm/llvm-project/pull/110486)

**Comment by LiqinWeng** [comment]
LGTM

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 2]
```diff
@@ -1383,6 +1385,15 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
                 VPValue *NewMask = GetNewMask(Red->getCondOp());
                 return new VPReductionEVLRecipe(*Red, EVL, NewMask);
               })
+              .Case<VPWidenSelectRecipe>([&](VPWidenSelectRecipe *Sel) {
+                SmallVector<VPValue *> Ops(Sel->operands());
+                Ops.push_back(&EVL);
+                return new VPWidenIntrinsicRecipe(
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1391]: What about cost estimation for this intrinsic? Do we rely on TTI here completely and no other cost tuning required?

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Looking at https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp#L1132, `vp_select` should match `SelectInst`'s cost.

There may be an edge-case where the condition is considered scalar (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L6631), but this shouldn't apply to `vp_select` and the VPlan cost would be more accurate. We may have to consider if EVL will be applied in the legacy cost computation if that indeed becomes an issue.

**Review by fhahn — COMMENTED** [review]
Update after https://github.com/llvm/llvm-project/pull/110486 merged

**Review by alexey-bataev — APPROVED** [review]

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -1042,6 +1041,14 @@ StringRef VPWidenIntrinsicRecipe::getIntrinsicName() const {
   return Intrinsic::getBaseName(VectorIntrinsicID);
 }
 
+bool VPWidenIntrinsicRecipe::onlyFirstLaneUsed(const VPValue *Op) const {
+  assert(is_contained(operands(), Op) && "Op must be an operand of the recipe");
+  // Vector predication intrinsics only demand the the first lane the last
+  // operand (the EVL operand).
+  return VPIntrinsic::isVPIntrinsic(VectorIntrinsicID) &&
+         Op == getOperand(getNumOperands() - 1);
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1049]: Meet an implementation divergence when rebasing #101641.
```
  return all_of(enumerate(operands()), [this, Op](const auto &I) {
    return Op != I.value() || 
           isVectorIntrinsicWithScalarOpAtArg(VectorIntrinsicID, I.index()) ||
           (VPIntrinsic::isVPIntrinsic(VectorIntrinsicID) && 
            I.index() == getNumOperands() - 1);
  });
```
Maybe we need to consider isVectorIntrinsicWithScalarOpAtArg?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 4]
```diff
@@ -79,7 +79,6 @@ bool VPRecipeBase::mayWriteToMemory() const {
     return !cast<VPWidenCallRecipe>(this)
                 ->getCalledScalarFunction()
                 ->onlyReadsMemory();
-  case VPWidenIntrinsicSC:
```
Comment (Mel-Chen) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:82]: Why remove this?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1049]: I don't see the changes in the current version of the PR I think? Would probably be good to improve (I assume scalar arg here means we only request the first lane). If there are unrelated tests that improve, would be good to prepare a separate patch

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 4 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:82]: Dropped by accident, should be restored in a4819bd46d8baebc3aaa8b38f78065de33593199, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1049]: Dropped the related changes while resolving conflicts, as the patch can still function even without considering `isVectorIntrinsicWithScalarOpAtArg`. So far, I haven't encountered any related cases.

**Pull request merged**